File: core/src/main/java/network/aika/activations/ConjunctiveActivation.java
package network.aika.activations;

import network.aika.Document;
import network.aika.bindingsignal.BindingSignal;
import network.aika.neurons.Neuron;
import network.aika.neurons.Synapse;
import network.aika.typedefs.ActivationDefinition;
import network.aika.bindingsignal.BSType;

import java.util.Map;
import java.util.NavigableMap;
import java.util.TreeMap;
import java.util.stream.Stream;


public class ConjunctiveActivation extends Activation {

    protected NavigableMap<Integer, Link> inputLinks = new TreeMap<>();

    public ConjunctiveActivation(
            ActivationDefinition t,
            Activation parent,
            Integer id,
            Neuron n,
            Document doc,
            Map<BSType, BindingSignal> bindingSignals
    ) {
        super(t, parent, id, n, doc, bindingSignals);
    }

    @Override
    public void linkIncoming(Activation excludedInputAct) {
        neuron
                .getInputSynapsesAsStream()
                .filter(s ->
                        s.isIncomingLinkingCandidate(getBindingSignals().keySet())
                )
                .forEach(s ->
                        linkIncoming(s, excludedInputAct)
                );
    }

    void linkIncoming(Synapse targetSyn, Activation excludedInputAct) {
        collectLinkingTargets(targetSyn.getInput(getModel())).stream()
                .filter(iAct -> iAct != excludedInputAct)
                .forEach(iAct ->
                        targetSyn.createLink(
                                iAct,
                                this
                        )
                );
    }

    @Override
    public void addInputLink(Link l) {
        Synapse syn = l.getSynapse();
        assert inputLinks.get(syn.getSynapseId()) == null;
        inputLinks.put(syn.getSynapseId(), l);
    }

    @Override
    public Link getInputLink(Activation iAct, int synapseId) {
        return inputLinks.get(synapseId);
    }

    @Override
    public Stream<Link> getInputLinks() {
        return inputLinks.values()
                .stream();
    }
}

File: src/network/conjunctive_activation.cpp
#include "network/conjunctive_activation.h"
#include "network/synapse.h"
#include "network/synapse_type.h"
#include "fields/relation.h"
#include "fields/rel_obj_iterator.h"
#include <cassert>

ConjunctiveActivation::ConjunctiveActivation(ActivationType* t, Activation* parent, int id, Neuron* n, Document* doc, std::map<BSType*, BindingSignal*> bindingSignals)
    : Activation(t, parent, id, n, doc, bindingSignals) {}

ConjunctiveActivation::~ConjunctiveActivation() {}

RelatedObjectIterable* ConjunctiveActivation::followManyRelation(Relation* rel) const {
    if (rel->getRelationLabel() == "INPUT") {
        // Convert inputLinks to a vector of Object*
        std::vector<Object*> objs;
        for (const auto& pair : inputLinks) {
            objs.push_back(static_cast<Object*>(pair.second));
        }
        return new VectorObjectIterable(objs);
    } else {
        // Use base class implementation for other relations
        return Activation::followManyRelation(rel);
    }
}

void ConjunctiveActivation::linkIncoming(Activation* excludedInputAct) {
    for (auto& s : neuron->getInputSynapsesAsStream()) {
        // Extract keys from the map (BSType* pointers)
        std::set<BSType*> bsKeys;
        for (const auto& pair : getBindingSignals()) {
            bsKeys.insert(pair.first);
        }
        if (static_cast<SynapseType*>(s->getType())->isIncomingLinkingCandidate(bsKeys)) {
            linkIncoming(s, excludedInputAct);
        }
    }
}

void ConjunctiveActivation::linkIncoming(Synapse* targetSyn, Activation* excludedInputAct) {
    for (auto& iAct : collectLinkingTargets(targetSyn->getInput(getModel()))) {
        if (iAct != excludedInputAct) {
            targetSyn->createLink(iAct, this);
        }
    }
}

void ConjunctiveActivation::addInputLink(Link* l) {
    Synapse* syn = l->getSynapse();
    assert(inputLinks.find(syn->getSynapseId()) == inputLinks.end());
    inputLinks[syn->getSynapseId()] = l;
}

std::vector<Link*> ConjunctiveActivation::getInputLinks() const {
    std::vector<Link*> result;
    for (const auto& pair : inputLinks) {
        result.push_back(pair.second);
    }
    return result;
} 